In [20]:
import itertools
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch_geometric.nn as gnn
import numpy as np
import pandas as pd
from torch_geometric.data import DataLoader
from torch_geometric.data import Data
from torch_geometric.data import InMemoryDataset
from torch_geometric.utils import negative_sampling
from torch_geometric.nn import GCNConv 
from torch.optim.lr_scheduler import StepLR
from sklearn.decomposition import TruncatedSVD
from sklearn.metrics import roc_auc_score, ndcg_score
from sklearn.metrics.pairwise import cosine_similarity
from tqdm import tqdm

In [2]:
class TwitterDataset(InMemoryDataset):
    def __init__(self, root, transform=None, pre_transform=None):
        super(TwitterDataset, self).__init__(root, transform, pre_transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return ['twitter_combined.txt', 'twitter_features.txt']

    @property
    def processed_file_names(self):
        return ['twitter_data.pt']

    def download(self):
        # In this case, you're not downloading any data,
        # but you could add code to download your data here if it's not local
        pass

    def process(self):
        # Load the graph data
        edges = pd.read_csv(self.raw_paths[0], delimiter=' ', header=None)
        features = pd.read_csv(self.raw_paths[1], delimiter=' ', header=None)

        # Create the graph
        edge_index = torch.tensor(edges.values, dtype=torch.long).t().contiguous()
        x = torch.tensor(features.values, dtype=torch.float)
    
        data = Data(x=x, edge_index=edge_index)

        data, slices = self.collate([data])
        torch.save((data, slices), self.processed_paths[0])
        
class GPlusDataset(InMemoryDataset):
    def __init__(self, root, transform=None, pre_transform=None):
        super(GPlusDataset, self).__init__(root, transform, pre_transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return ['gplus_combined.txt', 'gplus_features.txt']

    @property
    def processed_file_names(self):
        return ['gplus_data.pt']

    def download(self):
        # In this case, you're not downloading any data,
        # but you could add code to download your data here if it's not local
        pass

    def process(self):
        # Load the graph data
        edges = pd.read_csv(self.raw_paths[0], delimiter=' ', header=None)
        features = pd.read_csv(self.raw_paths[1], delimiter=' ', header=None)

        # Create the graph
        edge_index = torch.tensor(edges.values, dtype=torch.long).t().contiguous()
        x = torch.tensor(features.values, dtype=torch.float)
    
        data = Data(x=x, edge_index=edge_index)

        data, slices = self.collate([data])
        torch.save((data, slices), self.processed_paths[0])

In [3]:
class FacebookDataset(InMemoryDataset):
    def __init__(self, root, transform=None, pre_transform=None):
        super(FacebookDataset, self).__init__(root, transform, pre_transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return ['facebook_combined.txt', 'facebook_features.txt']

    @property
    def processed_file_names(self):
        return ['data.pt']

    def download(self):
        # In this case, you're not downloading any data,
        # but you could add code to download your data here if it's not local
        pass

    def process(self):
        # Load the graph data
        edges = pd.read_csv(self.raw_paths[0], delimiter=' ', header=None)
        features = pd.read_csv(self.raw_paths[1], delimiter=' ', header=None)

        # Create the graph
        edge_index = torch.tensor(edges.values, dtype=torch.long).t().contiguous()
        x = torch.tensor(features.values, dtype=torch.float)
    
        data = Data(x=x, edge_index=edge_index)

        data, slices = self.collate([data])
        torch.save((data, slices), self.processed_paths[0])

In [4]:
dataset = FacebookDataset(root='.')
data = dataset[0]

In [7]:
#Calculate Train Positives
# Define the percentage of edges to remove
test_percentage = 0.3

# Calculate the number of edges to remove
num_edges_to_remove = int(data.edge_index.shape[1] * test_percentage)

# Shuffle the edges
edge_indices = np.arange(data.edge_index.shape[1])
np.random.shuffle(edge_indices)

# Select the edges to keep
edges_to_keep = edge_indices[num_edges_to_remove:]

# Create a new graph with only the edges to keep(Train positives)
data_prime = Data(x=data.x, edge_index=data.edge_index[:, edges_to_keep])

In [9]:
#train pos 
train_pos=data_prime

In [11]:
#TEST POSITIVES
# Select the edges to remove (these will be your test positives)
edges_to_remove = edge_indices[:num_edges_to_remove]

# Create a new graph with only the edges to remove (Test positives)
test_pos = Data(x=data.x, edge_index=data.edge_index[:, edges_to_remove])

In [14]:
#Calculate Train negatives
positives_edges=data_prime.edge_index
# Number of negative samples to generate
num_neg_samples = data_prime.edge_index.size(1)
# Generate negative samples
negative_edge_index = negative_sampling(edge_index=data.edge_index, num_nodes=data.num_nodes, num_neg_samples=num_neg_samples)
#Train negatives
train_neg = Data(x=data.x, edge_index=negative_edge_index)

In [17]:
# TEST negatives
# Number of negative samples to generate for testing
num_test_neg_samples = test_pos.edge_index.size(1)

# Generate negative samples for testing
test_negative_edge_index = negative_sampling(edge_index=data.edge_index, num_nodes=data.num_nodes, num_neg_samples=num_test_neg_samples)

# Convert tensors to lists of tuples
train_pos_edges = [tuple(edge) for edge in train_pos.edge_index.t().tolist()]
test_pos_edges = [tuple(edge) for edge in test_pos.edge_index.t().tolist()]
test_negative_edges = [tuple(edge) for edge in test_negative_edge_index.t().tolist()]

# Ensure that these negative samples are not in the train_pos or test_pos sets
test_neg_edge_index = [edge for edge in test_negative_edges if edge not in train_pos_edges and edge not in test_pos_edges]

# Create the test_neg set
test_neg = Data(x=data.x, edge_index=torch.tensor(test_neg_edge_index).t().contiguous())

In [21]:
class GCN(torch.nn.Module):
    def __init__(self, num_features, hidden_size):
        super(GCN, self).__init__()
        self.conv1 = GCNConv(num_features, hidden_size)
        self.conv2 = GCNConv(hidden_size, hidden_size)

    def forward(self, edge_index, x):
        x = self.conv1(x, edge_index)
        x = torch.relu(x)
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index)
        return x

In [22]:
class TruncatedSVDModel(torch.nn.Module):
    def __init__(self, num_features, output_size):
        super(TruncatedSVDModel, self).__init__()
        self.svd = TruncatedSVD(n_components=output_size)
        self.fc = torch.nn.Linear(output_size, output_size)

    def forward(self, x):
        x_svd = self.svd.fit_transform(x.numpy())
        x_svd = torch.tensor(x_svd, dtype=torch.float)
        x_out = self.fc(x_svd)
        return x_out

In [23]:
class DotPredictor(torch.nn.Module):
    def forward(self, edge_index, h):
        # Get the embeddings of the source nodes and destination nodes
        source_node_embeddings = h[edge_index[0]]
        destination_node_embeddings = h[edge_index[1]]

        # Compute the dot product (score) between source and destination node embeddings
        scores = (source_node_embeddings * destination_node_embeddings).sum(dim=-1)

        return scores

In [36]:
def pipeline(model_name='GCN', hidden_size=64, epoch=100):
    
    def compute_loss(pos_score, neg_score):  # computes the loss based on binary cross entropy
        scores = torch.cat([pos_score, neg_score])
        labels = torch.cat([torch.ones(pos_score.shape[0]), torch.zeros(neg_score.shape[0])])
        return F.binary_cross_entropy_with_logits(scores, labels)

    def compute_auc(pos_score, neg_score):  # computes AUC (Area-Under-Curve) score
        scores = torch.cat([pos_score, neg_score]).numpy()
        labels = torch.cat(
            [torch.ones(pos_score.shape[0]), torch.zeros(neg_score.shape[0])]).numpy()
        return roc_auc_score(labels, scores)
       
    # hidden_size is the size of the hidden layer in the neural net
    if model_name == 'GCN':
        model = GCN(data_prime.num_features, hidden_size)
    elif model_name == 'TruncatedSVD':
        model = TruncatedSVDModel(data_prime.num_features, hidden_size)
        
    pred = DotPredictor()
    optimizer = torch.optim.SGD(itertools.chain(model.parameters(), pred.parameters()), lr=0.01, momentum=0.9)
    # Use a learning rate scheduler
    scheduler = StepLR(optimizer, step_size=500, gamma=0.5)
 

        # ----------- training -------------------------------- #
    train_g = data_prime
    for e in range(epoch):
        if model_name == 'GCN':
            h = model(train_g.edge_index, train_g.x)
        elif model_name == 'TruncatedSVD':
            h = model(train_g.x)  # get node embeddings

        # forward    
        pos_score = pred(train_pos.edge_index, h)
        neg_score = pred(train_neg.edge_index, h)
        loss = compute_loss(pos_score, neg_score)

        # backward
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()
        if e % 10 == 0:
            print('In epoch {}, loss: {}'.format(e, loss))

    # ----------- test and check results ---------------- #
    with torch.no_grad():
        pos_score = pred(test_pos.edge_index, h)
        neg_score = pred(test_neg.edge_index, h)
        auc=compute_auc(pos_score, neg_score)
        print('AUC', auc)  
               
    # Print model's state_dict
    print("Model's state_dict:")
    for param_tensor in model.state_dict():
        print(param_tensor, "\t", model.state_dict()[param_tensor].size())

    # Print optimizer's state_dict
    print("Optimizer's state_dict:")
    for var_name in optimizer.state_dict():
        print(var_name, "\t", optimizer.state_dict()[var_name])

    # Print scheduler's state_dict
    print("scheduler's state_dict:")
    for var_name in scheduler.state_dict():
        print(var_name, "\t", scheduler.state_dict()[var_name])
    
    torch.save({
            'epoch': epoch,        
            'epoch_rem': e,
            'model_state_dict': model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'scheduler_state_dict': scheduler.state_dict(),
            'loss': loss,
            
            }, './torch_model/model_'+model_name+'_'+str(epoch)+'.pt')
        
    return h  # return node embeddings

In [26]:
def generate_rec(h, user_id=0, k=10):
    # `h` represents the node embeddings, with shape [num_nodes, hidden_size]

    # generate a graph with (num_nodes - num_friends_of_user) edges
    # one end of the edge is user_id
    # the other end is a user that's NOT friends with user_id
    user_friends = set()
    user_neg_u, user_neg_v = [], []

    for n1, n2 in data.edge_index.t().tolist():   # get all friends of user_id
        if int(n1) == user_id:
            user_friends.add(int(n2))
        if int(n2) == user_id:
            user_friends.add(int(n1))

    num_nodes=data.x.shape[0]
    for i in range(num_nodes):  # generate "negative edges" for user_id
        if i != user_id and i not in user_friends:
            user_neg_u.append(user_id)
            user_neg_v.append(i)

    user_g = Data(x=data.x,edge_index=torch.tensor([user_neg_u, user_neg_v] ))

    pred = DotPredictor()

    # calculate the score of each user
    scores = []
    for i, score in enumerate(pred(user_g.edge_index, h)):
        rel=1 if ((user_id,i) in test_pos_edges) else 0
        scores.append((i, score,rel))

    # produce final ranked list
    scores.sort(key=lambda x: -x[1])

    # display results
    
    if (k !=0):
        print(f"List of 5 suggested friends for user {user_id}:")
    for i in range(k):
        print(f'- User {scores[i][0]}, score = {scores[i][1]}, rel = {scores[i][2]}')
    return scores[:10]


In [27]:
def calc_rec(h, user_id=0, k=10):
    user_friends = set()
    user_neg_u, user_neg_v = [], []

    for n1, n2 in data.edge_index.t().tolist():   # get all friends of user_id
        if int(n1) == user_id:
            user_friends.add(int(n2))
        if int(n2) == user_id:
            user_friends.add(int(n1))

    num_nodes=data.x.shape[0]
    for i in range(num_nodes):  # generate "negative edges" for user_id
        if i != user_id and i not in user_friends:
            user_neg_u.append(user_id)
            user_neg_v.append(i)

    user_g = Data(x=data.x,edge_index=torch.tensor([user_neg_u, user_neg_v] ))

    pred = DotPredictor()

    # calculate the score of each user
    scores = []
    for i, score in enumerate(pred(user_g.edge_index, h)):
        rel=1 if ((user_id,i) in test_pos_edges) else 0
        scores.append((i, score,rel))

    # produce final ranked list
    scores.sort(key=lambda x: -x[1])
    rel=[x[2] for x in scores]
    # display results
    
    return rel[:500]

In [28]:
def metrics(a,k=50):
    top_k = a[:k]
    hits=pd.DataFrame(0,index=[0],columns=top_k.columns)
    recipr=pd.DataFrame(0,index=[0],columns=top_k.columns)
    dcg=pd.DataFrame(0,index=[0],columns=top_k.columns)
    idcg=pd.DataFrame(0,index=[0],columns=top_k.columns)
    for i in list(top_k.columns):
        hits[i]=(top_k[i].sum()/k)
        recipr[i]=top_k.index[top_k[i] == 1].min()
        dcg[i]=0
        idcg[i]=0
        top_sort=top_k[i]
        top_sort=top_sort.sort_values(ascending=False)
        for j in range(0,k):
            dcg[i]+=top_k[i].iloc[j]/np.log2(j+1+1)
            idcg[i]+=top_sort.iloc[j]/np.log2(j+1+1)

    recipr=recipr.replace(np.nan,-1) +1
    recipr = recipr.loc[:, (recipr != 0).any()]
    ndcg=dcg/idcg
    ndcg=ndcg.T.replace(np.nan,0)

    ndcg_k=np.mean(ndcg)
    mrr=(1/recipr.T).sum()[0]/len(a.columns)
    hits_at_k=np.mean(hits)

    print('Hits@'+str(k)+':', hits_at_k)
    print('NDCG@'+str(k)+':', ndcg_k)
    print('MRR:', mrr)
    
    return hits_at_k, ndcg_k, mrr


In [29]:
h = pipeline("GCN",epoch=1000)

/home/vkaraiskos/.local/lib/python3.8/site-packages/torch/autograd/__init__.py:251: UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 11040). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:108.)
  Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass


In epoch 0, loss: 0.6360924243927002
In epoch 10, loss: 0.6108479499816895
In epoch 20, loss: 0.576044499874115
In epoch 30, loss: 0.5548127293586731
In epoch 40, loss: 0.5355092883110046
In epoch 50, loss: 0.522758960723877
In epoch 60, loss: 0.5067281126976013
In epoch 70, loss: 0.5022969245910645
In epoch 80, loss: 0.49420297145843506
In epoch 90, loss: 0.48290392756462097
In epoch 100, loss: 0.48026585578918457
In epoch 110, loss: 0.47341370582580566
In epoch 120, loss: 0.46738937497138977
In epoch 130, loss: 0.46773308515548706
In epoch 140, loss: 0.4603409171104431
In epoch 150, loss: 0.46276819705963135
In epoch 160, loss: 0.4598846435546875
In epoch 170, loss: 0.4583229124546051
In epoch 180, loss: 0.4569869041442871
In epoch 190, loss: 0.45694300532341003
In epoch 200, loss: 0.4540381133556366
In epoch 210, loss: 0.45437711477279663
In epoch 220, loss: 0.45420098304748535
In epoch 230, loss: 0.4507874846458435
In epoch 240, loss: 0.4496913254261017
In epoch 250, loss: 0.452769

In [31]:
target_users = np.unique(np.array(test_pos_edges).flatten())
rankings={}
with tqdm(total=len(target_users), position=0, leave=True) as pbar:
    for user in target_users:         
        rankings[user]=calc_rec(h,user)
        pbar.update()



100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3820/3820 [3:24:27<00:00,  3.21s/it]


In [37]:
h2 = pipeline("TruncatedSVD",epoch=1000)

In epoch 0, loss: 0.8319059014320374
In epoch 10, loss: 0.6391419768333435
In epoch 20, loss: 0.6264968514442444
In epoch 30, loss: 0.6207771897315979
In epoch 40, loss: 0.6163594722747803
In epoch 50, loss: 0.6100722551345825
In epoch 60, loss: 0.6041662096977234
In epoch 70, loss: 0.6012915372848511
In epoch 80, loss: 0.5962530970573425
In epoch 90, loss: 0.591964840888977
In epoch 100, loss: 0.5883418321609497
In epoch 110, loss: 0.5851919651031494
In epoch 120, loss: 0.5824350118637085
In epoch 130, loss: 0.5787000060081482
In epoch 140, loss: 0.5758125185966492
In epoch 150, loss: 0.5737694501876831
In epoch 160, loss: 0.5708563327789307
In epoch 170, loss: 0.567463219165802
In epoch 180, loss: 0.5653952360153198
In epoch 190, loss: 0.5638371109962463
In epoch 200, loss: 0.5628612637519836
In epoch 210, loss: 0.5612289309501648
In epoch 220, loss: 0.5587643384933472
In epoch 230, loss: 0.5574308037757874
In epoch 240, loss: 0.555939257144928
In epoch 250, loss: 0.5546720027923584


In [39]:
rankings2={}
with tqdm(total=len(target_users), position=0, leave=True) as pbar:
    for user in target_users:              
        rankings2[user]=calc_rec(h2,user)
        pbar.update()

100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3820/3820 [3:25:34<00:00,  3.23s/it]


In [32]:
a=pd.DataFrame(rankings)
#all users
metrics(a,k=50)

Hits@50: 0.009544502617801048
NDCG@50: 0.08679938917996073
MRR: 0.03602528725553057


(0.009544502617801048, 0.08679938917996073, 0.03602528725553057)

In [33]:
metrics(a,k=5)

Hits@5: 0.009633507853403141
NDCG@5: 0.027250880903910743
MRR: 0.021156195462478183


(0.009633507853403141, 0.027250880903910743, 0.021156195462478183)

In [34]:
a.to_csv('dot_product_gcn_all.csv', index=False) 
a.to_pickle('dot_product_gcn_all.pkl')
torch.save(h, 'node_embeddings_gcn.pt')

In [40]:
a2=pd.DataFrame(rankings2)
#all users
metrics(a2,k=50)

Hits@50: 0.004816753926701571
NDCG@50: 0.04478218425877151
MRR: 0.017061470995251078


(0.004816753926701571, 0.04478218425877151, 0.017061470995251078)

In [41]:
metrics(a2,k=5)

Hits@5: 0.004397905759162304
NDCG@5: 0.012585534683067784
MRR: 0.009694589877835953


(0.004397905759162304, 0.012585534683067784, 0.009694589877835953)

In [42]:
a2.to_csv('dot_product_TruncatedSVD_all.csv', index=False) 
a2.to_pickle('dot_product_TruncatedSVD_all.pkl')
torch.save(h2, 'node_embeddings_TruncatedSVD.pt')

|model|AUC|hits@5|NDCG@5| MRR  |hits@50|NDCG@50|MRR|
|-----|---|---|------|------|-------|-------|---|
| GCN |0.9596|0.0096|0.0273|0.0212| 0.0095|0.0868 |0.036  |
|tSVD |0.8556|0.0043|0.0126|0.0097| 0.0048|0.0448 |0.017 |

In [91]:
# Load if unloaded and get recommendations
if 'h' not in globals() or 'h' not in locals():
    h=torch.load('node_embeddings_gcn.pt')
if 'a' not in globals() or 'a' not in locals():
    a = pd.read_pickle('dot_product_gcn_all.pkl')
if 'h2' not in globals() or 'h2' not in locals():
    h=torch.load('node_embeddings_TruncatedSVD.pt')
if 'a2' not in globals() or 'a2' not in locals():
    a = pd.read_pickle('dot_product_TruncatedSVD_all.pkl')

In [92]:
sc=generate_rec(h, user_id=40, k=5) #k only for visualization here

List of 5 suggested friends for user 40:
- User 232, score = 2.4407334327697754, rel = 1
- User 107, score = 2.2264106273651123, rel = 0
- User 225, score = 2.206943988800049, rel = 0
- User 170, score = 2.0902135372161865, rel = 0
- User 212, score = 2.015990972518921, rel = 1


In [93]:
sc2=generate_rec(h2, user_id=40, k=5)

List of 5 suggested friends for user 40:
- User 153, score = 3.9653868675231934, rel = 0
- User 2238, score = 3.624063014984131, rel = 0
- User 2090, score = 3.6001768112182617, rel = 0
- User 1898, score = 3.505584239959717, rel = 0
- User 210, score = 3.4285125732421875, rel = 0
